### PR TITLE
Update Sparkle Motion to state at Burning Man 2022

### PR DESCRIPTION
**Pull request by bencbartlett**
Added a shader that assigns each panel an ID and cycles through 4 colors. The hue can evolve over time at a slidable speed. (Also some other bugfixes.)

More to come!

Demo: (pls excuse the slow refresh rate from the simulator; the colors will change on the beat)

https://user-images.githubusercontent.com/6800253/184024025-6252327d-1daf-4c4e-8d44-7af636a84431.mov

### CODE REVIEW

**Comment by nzoschke** [comment]
Testing this here, will report back!

**Comment by nzoschke** [comment]
These are sick!

Screen Shot 2022-08-16 at 5 45 57 PM

File: src/commonMain/resources/shader-libraries/built-in/libraries/color_functions.glsl [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// Some commonly used color functions
+// Ben Bartlett
+
+/* Convert HSV values ot RGB values
+   Hue: cycles over range betwee 0.0 and 1.0; modulo 1
+   Saturaton: 0-1
+   Value: 0-1
+   */
+vec3 hsv2rgb(in vec3 c){
```
Comment (nzoschke) [inline src/commonMain/resources/shader-libraries/built-in/libraries/color_functions.glsl:9]: the comments are very useful. Are these functions available to call in any shader then?

File: src/commonMain/resources/shader-libraries/built-in/libraries/color_functions.glsl [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// Some commonly used color functions
+// Ben Bartlett
+
+/* Convert HSV values ot RGB values
```
Comment (nzoschke) [inline src/commonMain/resources/shader-libraries/built-in/libraries/color_functions.glsl:4]: typo, `ot` -> `to`

File: src/commonMain/resources/shader-libraries/built-in/shaders/BAAAHS_ColorPulse.glsl [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+struct BeatInfo {
+	float beat;
+	float bpm;
+	float intensity;
+	float confidence;
+};
+uniform BeatInfo beatInfo; // @@baaahs.BeatLink:BeatInfo
+
+/* Returns a value which cycles from 0 to 1 for each beat, with most of the increase occuring near the beat */
+float beatIntegral() {
```
Comment (nzoschke) [inline src/commonMain/resources/shader-libraries/built-in/shaders/BAAAHS_ColorPulse.glsl:20]: can you use these from the built-in libraries instead?

File: src/commonMain/resources/shader-libraries/built-in/libraries/color_functions.glsl [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// Some commonly used color functions
```
Comment (nzoschke) [inline src/commonMain/resources/shader-libraries/built-in/libraries/color_functions.glsl:1]: I've seen and used the `rgb2` counterparts, should we add those here too?

File: src/commonMain/resources/shader-libraries/built-in/shaders/BAAAHS_GeometricColorPulse.glsl [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	float riseTime = 0.3;
+	float peakTime = 0.2;
+	return smoothstep(0., 1., t / riseTime) - smoothstep(0., 1., (t - peakTime) / riseTime);
+}
+
+// @return color
+// @param uvIn uv-coordinate
+vec4 main(vec2 uvIn) {
+
+	int panelID = getPanelID(fixtureInfo);
```
Comment (nzoschke) [inline src/commonMain/resources/shader-libraries/built-in/shaders/BAAAHS_GeometricColorPulse.glsl:64]: for my education only, how do `uvIn` and `fixtureInfo` interact?

Comment (nzoschke) [reply src/commonMain/resources/shader-libraries/built-in/shaders/BAAAHS_GeometricColorPulse.glsl:64]: discussed on a call, its magic of sorts

File: src/commonMain/resources/shader-libraries/built-in/shaders/BAAAHS_ColorPulse.glsl (hunk 3 above)
Comment (nzoschke) [reply src/commonMain/resources/shader-libraries/built-in/shaders/BAAAHS_ColorPulse.glsl:20]: from a call, no there is no import, so copy/paste from the library

**Comment by bencbartlett** [comment]
@xian I ported over all of the additional changes I made on playa at Burning Man this year and added the new show template. I think this should be the most up-to-date version now

**Pull request merged**